### PR TITLE
Add new event commit status creation and webhook implementation

### LINKED ISSUE #20749: Create a notification on status update failures
### Feature Description

It would be nice to get a notification in GItea when an external service reports a failed status job back. For example a CI service could report a failed job and Gitea could directly show it as notification. 

### Screenshots

_No response_

### PR DESCRIPTION
**Pull request by lunny**
This PR introduces a new event which is similar as Github's. When a new commit status submitted, the event will be trigged. That means, now we can receive all feedback from CI/CD system in webhooks or other notify systems.

ref: https://docs.github.com/en/webhooks/webhook-events-and-payloads#status

Fix #20749

### CODE REVIEW

**Comment by anbraten** [comment]
Looks like a nice idea to have a webhook event for this. However the original idea behind the issue was rather to get an internal notification (shown when clicking on the bell in the UI) for failed PR status reports. So a CI like gitea actions would send a PR status update and if that is a failed one it would create a new notification entry in the list.

**Comment by denyskon** [comment]
@lunny Could you please explain this change? I don't understand which action is triggered by this method in the end......

**Comment by lunny** [comment]
> @lunny Could you please explain this change? I don't understand which action is triggered by this method in the end......

This event is triggered by a new commit status submitted to Gitea. Currently only webhook notfier implement it so then when a commit status has been submitted, a webhook will be triggered.

**Comment by denyskon** [comment]
@lunny I don't manage to trigger a webhook. Do I misunderstand you? I run a workflow on a repo, which succeeds or fails. I have a Telegram webhook which triggers on commit, but not on workflow success/failure

**Comment by lunny** [comment]
> @lunny I don't manage to trigger a webhook. Do I misunderstand you? I run a workflow on a repo, which succeeds or fails. I have a Telegram webhook which triggers on commit, but not on workflow success/failure

There is no CI/CD success/failure message directly. But we can use commit status success/failure with context and commitid to know whether/which the task success/failure.

**Comment by denyskon** [comment]
Yes, but do I have to activate it separately somewhere to also get commit status over webhook?
I only get the new commit message

**Comment by denyskon** [comment]
Ping :) Would really like to be able to review this before 1.22. Do I misunderstand something, or does it currently not fully work?

File: modules/structs/hook.go [hunk 1]
```diff
@@ -494,3 +487,27 @@ type PackagePayload struct {
...
+// CommitStatusPayload represents a payload information of commit status event.
+type CommitStatusPayload struct {
+	// TODO: add Branches
+	Commit  *PayloadCommit `json:"commit"`
+	Context string         `json:"context"`
+	// swagger:strfmt date-time
+	CreatedAt   time.Time `json:"created_at"`
+	Description string    `json:"description"`
+	ID          int64     `json:"id"`
+	// Name        string         `json:"name"`
```
Comment (delvh) [inline modules/structs/hook.go]: ?

File: modules/structs/hook.go [hunk 2]
```diff
@@ -494,3 +487,27 @@ type PackagePayload struct {
 func (p *PackagePayload) JSONPayload() ([]byte, error) {
 	return json.MarshalIndent(p, "", "  ")
 }
+
+// CommitStatusPayload represents a payload information of commit status event.
+type CommitStatusPayload struct {
+	// TODO: add Branches
```
Comment (delvh) [inline modules/structs/hook.go]: I am not quite sure what you mean by that.

File: services/repository/files/commit.go [hunk 3]
```diff
@@ -5,13 +5,65 @@ package files
...
+	repo_module "code.gitea.io/gitea/modules/repository"
 	"code.gitea.io/gitea/modules/structs"
+	"code.gitea.io/gitea/services/automerge"
+	"code.gitea.io/gitea/services/notify"
 )
 
+// CreateCommitStatus creates a new CommitStatus given a bunch of parameters
+// NOTE: All text-values will be trimmed from whitespaces.
+// Requires: Repo, Creator, SHA
+func CreateCommitStatus(ctx context.Context, repo *repo_model.Repository, creator *user_model.User, sha string, status *git_model.CommitStatus) error {
```
Comment (wolfogre) [inline services/repository/files/commit.go]: I don't understand. I don't see any usage of this function, how it works?

Comment (wolfogre) [reply services/repository/files/commit.go]: I believe it's related to #29444.

**Comment by denyskon** [comment]
@lunny I still am not able to trigger a webhook notification. What do I need to do to be able to test it?

File: services/repository/files/commit.go (hunk 3 above)
Comment (lunny) [reply services/repository/files/commit.go]: [9c24434](https://github.com/go-gitea/gitea/pull/27151/commits/9c24434d3eefbb6130ca2c43813121d3d2c38173)

**Comment by lunny** [comment]
> @lunny I still am not able to trigger a webhook notification. What do I need to do to be able to test it?

Sorry, I think it's not finished when you tested it. I think it's ready now.

**Comment by denyskon** [comment]
@lunny Still no success. If the feature is working as intended for you, could you please provide some instructions how to test it properly?

File: modules/structs/hook.go (hunk 1 above)
Comment (delvh) [reply modules/structs/hook.go]: Still, why is the `Name` commented out?

File: modules/structs/hook.go [hunk 4]
```diff
@@ -494,3 +487,27 @@ type PackagePayload struct {
...
+	Repo      *Repository `json:"repository"`
+	Sender    *User       `json:"sender"`
+	SHA       string      `json:"sha"`
+	State     string      `json:"state"`
+	TargetURL string      `json:"target_url"`
+	// swagger:strfmt date-time
+	UpdatedAt *time.Time `json:"updated_at"`
+}
+
+// JSONPayload FIXME
```
Comment (delvh) [inline modules/structs/hook.go]: What needs to be fixed? It looks complete to me...

File: modules/structs/hook.go (hunk 1 above)
Comment (lunny) [reply modules/structs/hook.go]: 9363b99f1ba24e3bbf0b34da44cba1025bfdabeb

File: modules/structs/hook.go (hunk 4 above)
Comment (lunny) [reply modules/structs/hook.go]: Removed.

**Review by wolfogre — APPROVED** [review]

**Review by yp05327 — APPROVED** [review]

**Pull request merged**